### PR TITLE
Revamp the expressions/paths

### PR DESCRIPTION
**Pull request by GuillaumeDerval**
- Remove the concept of boolean expressions (everything is an expression now)
- Remove var_or_param
- Add Path for when we reallllly want the path to something rather than an expression
- Dot calls (`a.b`), array indexing (`a[b]`) and functions calls (`a(b)`) are now part of the expression AST

### CODE REVIEW

**Comment by GuillaumeDerval** [comment]
(don't merge this now)

File: tests/instances/ok/boolean_checks.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+#NODE H
+    #PARAMETERS
+        a = 2;
+        b = (a == 2);
+        c = {b, a == 3};
+    #VARIABLES
+        internal : x;
+    #CONSTRAINTS
+        CUSTOMFUNC(a == 2);
```
Comment (T-3B) [inline tests/instances/ok/boolean_checks.txt:11]: For this to pass semantic checking, `CUSTOMFUNC` should be defined

Comment (GuillaumeDerval) [reply tests/instances/ok/boolean_checks.txt:11]: For now, the tests are mainly for parsing; we indeed need to separate "things that should parse, but are not semantically sound" from "things that should parse and are semantically sound" but that's not the purpose of this PR

**Review by MiftariB — APPROVED** [review]

**Comment by MiftariB** [comment]
LGTM

**Pull request merged**